### PR TITLE
ipv6: disable IPv6-only shared network with VR

### LINKED ISSUE #5640: NPE while deploy vm on a IPv6-only shared network
Got a NPE
```
java.lang.NullPointerException
        at com.cloud.network.Networks$BroadcastDomainType.getSchemeValue(Networks.java:177)
        at com.cloud.network.Networks$BroadcastDomainType.getValue(Networks.java:231)
        at com.cloud.hypervisor.kvm.resource.BridgeVifDriver.plug(BridgeVifDriver.java:209)
        at com.cloud.hypervisor.kvm.resource.LibvirtComputingResource.createVif(LibvirtComputingResource.java:3015)
        at com.cloud.hypervisor.kvm.resource.LibvirtComputingResource.createVifs(LibvirtComputingResource.java:2710)
        at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStartCommandWrapper.execute(LibvirtStartCommandWrapper.java:81)
        at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStartCommandWrapper.execute(LibvirtStartCommandWrapper.java:46)
        at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtRequestWrapper.execute(LibvirtRequestWrapper.java:78)
        at com.cloud.hypervisor.kvm.resource.LibvirtComputingResource.executeRequest(LibvirtComputingResource.java:1768)
```

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
management server
~~~

##### CLOUDSTACK VERSION

~~~
4.16.0.0
~~~

##### CONFIGURATION

advanced zone

##### OS / ENVIRONMENT

Mgmt server: Ubuntu 20.04
Hypervisor: Ubuntu 20.04

##### STEPS TO REPRODUCE

~~~
1. create a shared network with only ipv6
2. deploy a vm
~~~

##### EXPECTED RESULTS

~~~
succeed
~~~

##### ACTUAL RESULTS

~~~
Got exception, see the description
~~~

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #5640 by disabling IPv6-only Shared network with VR provider.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java [hunk 1]
```diff
@@ -1252,8 +1252,8 @@ public Network createGuestNetwork(CreateNetworkCmd cmd) throws InsufficientCapac
             }
         }
 
-        boolean ipv4 = true, ipv6 = false;
-        if (startIP != null) {
+        boolean ipv4 = false, ipv6 = false;
+        if (org.apache.commons.lang3.StringUtils.isNoneBlank(gateway, netmask)) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/network/NetworkServiceImpl.java:1256]: @weizhouapache, recently we adopted `org.apache.commons.lang3.StringUtils` as the default string lib. A checkstyle validation was added to prohibit imports of `org.apache.commons.lang.StringUtils` and `com.google.common.base.Strings`, however it does not cover `import static`. The files that used `import static` were modified to remove the `import static` and use the object `StringUtils` from `lang3`, however, this file went unnoticed. 

Could you remove those `import static` (`import static org.apache.commons.lang.StringUtils.isBlank;` and `import static org.apache.commons.lang.StringUtils.isNotBlank;`) from the begin of the file, import the `org.apache.commons.lang3.StringUtils` and use it instead?

I observed that `com.cloud.utils.StringUtils` is already imported and would conflict with `org.apache.commons.lang3.StringUtils`. However, `com.cloud.utils.StringUtils` is used only twice and `org.apache.commons.lang3.StringUtils` will be used several times, therefore, I'd recommend to import `org.apache.commons.lang3.StringUtils` and to use the full class name of `com.cloud.utils.StringUtils`.

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java [hunk 2]
```diff
@@ -1338,6 +1334,10 @@ public Network createGuestNetwork(CreateNetworkCmd cmd) throws InsufficientCapac
             if(isBlank(zone.getIp6Dns1()) && isBlank(zone.getIp6Dns2())) {
                 throw new InvalidParameterValueException("Can only create IPv6 network if the zone has IPv6 DNS! Please configure the zone IPv6 DNS1 and/or IPv6 DNS2.");
             }
+
+            if (!ipv4 && ntwkOff.getGuestType() == GuestType.Shared && _networkModel.isProviderForNetworkOffering(Provider.VirtualRouter, networkOfferingId)) {
+                throw new InvalidParameterValueException("Currently IPv6-only Shared network with Virtual Router provider is not supported");
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/network/NetworkServiceImpl.java:1339]:
```suggestion
                throw new InvalidParameterValueException("Currently IPv6-only Shared network with Virtual Router provider is not supported.");
```

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 3]
```diff
@@ -2642,7 +2642,7 @@ private Network createGuestNetwork(final long networkOfferingId, final String na
                 && !_networkModel.areServicesSupportedByNetworkOffering(ntwkOff.getId(), Service.SourceNat)));
         if (cidr == null && ip6Cidr == null && cidrRequired) {
             if (ntwkOff.getGuestType() == GuestType.Shared) {
-                throw new InvalidParameterValueException("StartIp/endIp/gateway/netmask are required when create network of" + " type " + Network.GuestType.Shared);
+                throw new InvalidParameterValueException("gateway/netmask are required when create network of" + " type " + Network.GuestType.Shared);
```
Comment (GutoVeronezi) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:2645]:
```suggestion
                throw new InvalidParameterValueException(String.format("Gateway/netmask are required when creating %s networks.", Network.GuestType.Shared));
```

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java (hunk 1 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/network/NetworkServiceImpl.java:1256]: > @weizhouapache, recently we adopted `org.apache.commons.lang3.StringUtils` as the default string lib. A checkstyle validation was added to prohibit imports of `org.apache.commons.lang.StringUtils` and `com.google.common.base.Strings`, however it does not cover `import static`. The files that used `import static` were modified to remove the `import static` and use the object `StringUtils` from `lang3`, however, this file went unnoticed.
> 
> Could you remove those `import static` (`import static org.apache.commons.lang.StringUtils.isBlank;` and `import static org.apache.commons.lang.StringUtils.isNotBlank;`) from the begin of the file, import the `org.apache.commons.lang3.StringUtils` and use it instead?
> 
> I observed that `com.cloud.utils.StringUtils` is already imported and would conflict with `org.apache.commons.lang3.StringUtils`. However, `com.cloud.utils.StringUtils` is used only twice and `org.apache.commons.lang3.StringUtils` will be used several times, therefore, I'd recommend to import `org.apache.commons.lang3.StringUtils` and to use the full class name of `com.cloud.utils.StringUtils`.



@GutoVeronezi thanks for your comment. it is good #5386 has been merged into main branch. 
However 4.16 does not have it. com.cloud.utils.StringUtils is still imported.

https://github.com/apache/cloudstack/blob/4.16/server/src/main/java/com/cloud/network/NetworkServiceImpl.java#L175

**Comment by weizhouapache** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java (hunk 1 above)
Comment (GutoVeronezi) [reply server/src/main/java/com/cloud/network/NetworkServiceImpl.java:1256]: @weizhouapache, you're right. I didn't noticed it was based in 4.16. Ignore this comment. =3

**Comment by weizhouapache** [comment]
@blueorangutan test

**Review by GutoVeronezi — APPROVED** [review]
CLGTM

**Review by rohityadavcloud — APPROVED** [review]
LGTM, needs testing (if done/satisfied, pl merge)

**Comment by sureshanaparti** [comment]
@blueorangutan test keepEnv

**Review by sureshanaparti — APPROVED** [review]
Verified the changes. LGTM

**Comment by weizhouapache** [comment]
> Verified the changes. LGTM

Great, thanks for testing

**Pull request merged**